File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.rs
impl Solution {
    pub fn max_profit(k: i32, prices: Vec<i32>) -> i32 {
        let n = prices.len();
        let k = k as usize;
        if n == 0 || k == 0 {
            return 0;
        }

        let mut buy = vec![i32::MIN; k + 1];
        let mut sell = vec![0; k + 1];

        for &price in prices.iter() {
            for j in 1..=k {
                buy[j] = buy[j].max(sell[j - 1] - price);
                sell[j] = sell[j].max(buy[j] + price);
            }
        }

        sell[k]
    }
}

File: 0188-best-time-to-buy-and-sell-stock-iv/0188-best-time-to-buy-and-sell-stock-iv.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int n = prices.size();
        if (k >= n / 2) {
            int profit = 0;
            for (int i = 1; i < n; i++)
                if (prices[i] > prices[i-1]) profit += prices[i] - prices[i-1];
            return profit;
        }
        
        vector<int> buy(k, INT_MIN), sell(k, 0);
        
        for (int price : prices) {
            for (int i = 0; i < k; i++) {
                buy[i] = max(buy[i], (i == 0 ? 0 : sell[i-1]) - price);
                sell[i] = max(sell[i], buy[i] + price);
            }
        }
        
        return sell[k-1];
    }
};